libvips/foreign/openslideconnection.c

```diff
@@ -0,0 +1,252 @@
+{
+	VipsOpenslideConnection *oldest;
+
+	oldest = NULL;
+	g_hash_table_foreach(vips_openslideconnection_cache,
+		vips_openslideconnection_oldest_cb, &oldest);
+
+	return oldest;
+}
+
+static void
+vips_openslideconnection_trim(void)
+{
+	VipsOpenslideConnection *oldest;
+
+	while (vips_openslideconnection_unused() > OPENSLIDECONNECTION &&
+		(oldest = vips_openslideconnection_oldest()))
+		vips_openslideconnection_free(oldest);
+}
+
+static void
+vips_openslideconnection_unref(VipsOpenslideConnection *connection)
+{
+	gboolean trim;
+
+	g_mutex_lock(&connection->lock);
+
+	g_assert(connection->ref_count > 0);
+
+	connection->ref_count -= 1;
+	trim = connection->ref_count == 0;
+
+	g_mutex_unlock(&connection->lock);
+
+	if (trim)
+		vips_openslideconnection_trim();
+}
+
+static void
+vips_openslideconnection_ref(VipsOpenslideConnection *connection)
+{
+	g_mutex_lock(&connection->lock);
+
+	g_assert(connection->ref_count >= 0);
+
+	connection->ref_count += 1;
+
+	g_mutex_unlock(&connection->lock);
+}
+
+static VipsOpenslideConnection *
+vips_openslideconnection_new(const char *filename)
+{
+	VipsOpenslideConnection *connection;
+
+	connection = g_new0(VipsOpenslideConnection, 1);
+	connection->filename = g_strdup(filename);
+
+	g_assert(!g_hash_table_lookup(vips_openslideconnection_cache, filename));
+
+	g_hash_table_insert(vips_openslideconnection_cache,
+		connection->filename, connection);
+
+	return connection;
+}
+
+static void
+vips_openslideconnection_touch(VipsOpenslideConnection *connection)
+{
+	static int time = 0;
+
+	g_mutex_lock(&connection->lock);
+
+	connection->time = time++;
+
+	g_mutex_unlock(&connection->lock);
+}
+
+openslide_t *
+vips__openslideconnection_open(const char *filename)
+{
+	g_mutex_lock(&vips_openslideconnection_lock);
+
+	if (!vips_openslideconnection_cache)
+		vips_openslideconnection_cache =
+			g_hash_table_new(g_str_hash, g_str_equal);
+
+	VipsOpenslideConnection *connection;
+
+	connection = g_hash_table_lookup(vips_openslideconnection_cache, filename);
+
+	if (!connection)
+		connection = vips_openslideconnection_new(filename);
+
+	vips_openslideconnection_ref(connection);
+	vips_openslideconnection_touch(connection);
+
+	g_mutex_unlock(&vips_openslideconnection_lock);
+
+	g_mutex_lock(&connection->lock);
+
+	/* We do the open outside the main lock (just in the connection lock)
+	 * since it can take many seconds for some slides.
+	 */
+	if (!connection->osr)
+		connection->osr = openslide_open(connection->filename);
+
+	openslide_t *osr = connection->osr;
+
+	g_mutex_unlock(&connection->lock);
+
+	return osr;
+}
+
+void
+vips__openslideconnection_close(const char *filename)
+{
+	g_mutex_lock(&vips_openslideconnection_lock);
+
+	VipsOpenslideConnection *connection;
+	connection = g_hash_table_lookup(vips_openslideconnection_cache, filename);
+
+	/* We unref in the main lock since openslide_close() is always relatively
+	 * quick.
+	 */
+	if (connection)
+		vips_openslideconnection_unref(connection);
```

Comment:
If we're called on an `openslide_t` in error state, we should detect that and stop handing that `openslide_t` out to new requestors, so errors don't become unhelpfully sticky.  Unfortunately the fact that we currently only take a `filename` argument makes this difficult.  If it were me, I'd have this function take a `VipsOpenslideConnection`, have `vips__openslideconnection_open` return that in an out-arg, and remove the `VipsOpenslideConnection` from the hash table if the `openslide_t` is closed from an error state.

Reply:
That's a good point. I think dropping the error state connection from cache on ref-count == 0 (to force a reopen) is probably the best solution.